Exception: empty collection can't be reduced

This happened when I deleted a certificate secret - to add more subdomains to it - and expected that `kubernetes-letsencrypt` would recreate it:

``` 
Exception in thread "Thread-26398" java.lang.UnsupportedOperationException: java.lang.UnsupportedOperationException: Empty collection can't be reduced.
	at sun.reflect.GeneratedConstructorAccessor54.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:593)
	at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:677)
	at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:735)
	at java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:160)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:583)
	at in.tazj.k8s.letsencrypt.acme.CertificateRequestHandler.requestCertificate(CertificateRequestHandler.kt:41)
	at in.tazj.k8s.letsencrypt.kubernetes.ServiceManager.handleCertificateRequest(ServiceManager.kt:64)
	at in.tazj.k8s.letsencrypt.kubernetes.ServiceManager.access$handleCertificateRequest(ServiceManager.kt:20)
	at in.tazj.k8s.letsencrypt.kubernetes.ServiceManager$reconcileService$1.run(ServiceManager.kt:45)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.UnsupportedOperationException: Empty collection can't be reduced.
	at in.tazj.k8s.letsencrypt.acme.CloudDnsResponder.findMatchingZone(CloudDnsResponder.kt:123)
	at in.tazj.k8s.letsencrypt.acme.CloudDnsResponder.updateCloudDnsRecord(CloudDnsResponder.kt:55)
	at in.tazj.k8s.letsencrypt.acme.CloudDnsResponder.addChallengeRecord(CloudDnsResponder.kt:26)
	at in.tazj.k8s.letsencrypt.acme.CertificateRequestHandler.prepareDnsChallenge(CertificateRequestHandler.kt:176)
	at in.tazj.k8s.letsencrypt.acme.CertificateRequestHandler.authorizeDomain(CertificateRequestHandler.kt:77)
	at in.tazj.k8s.letsencrypt.acme.CertificateRequestHandler.access$authorizeDomain(CertificateRequestHandler.kt:27)
	at in.tazj.k8s.letsencrypt.acme.CertificateRequestHandler$requestCertificate$1.accept(CertificateRequestHandler.kt:41)
	at in.tazj.k8s.letsencrypt.acme.CertificateRequestHandler$requestCertificate$1.accept(CertificateRequestHandler.kt:27)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:291)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) 
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -7,4 +7,4 @@ services:
   - docker
 before_install:
   - docker --version
-  - 'if [ "$TRAVIS_PULL_REQUEST" = "false" ]; then docker login -e="$DOCKER_EMAIL" -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD"; fi'
+  - 'if [ "$TRAVIS_PULL_REQUEST" = "false" ]; then docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD"; fi'

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -10,7 +10,7 @@ buildscript {
 
     dependencies {
         classpath(kotlinModule("gradle-plugin"))
-        classpath("com.bmuschko:gradle-docker-plugin:3.0.9")
+        classpath("com.bmuschko:gradle-docker-plugin:3.2.2")
         classpath("com.github.ben-manes:gradle-versions-plugin:0.15.0")
     }
 }

--- a/src/main/kotlin/in/tazj/k8s/letsencrypt/acme/CloudDnsResponder.kt
+++ b/src/main/kotlin/in/tazj/k8s/letsencrypt/acme/CloudDnsResponder.kt
@@ -92,7 +92,13 @@ class CloudDnsResponder(private val dns: Dns) : DnsResponder {
      * Determine the most specific matching zone from available Cloud DNS zones.
      */
     fun findMatchingZone(recordName: String): Option<Zone> {
-        return fetchMatchingZones(recordName)
+        val zones = fetchMatchingZones(recordName)
+
+        if (zones.isEmpty()) {
+            return Option.None
+        }
+
+        return zones
                 .reduce { acc, zone ->
                     if (zone.dnsName.length > acc.dnsName.length) {
                         zone

--- a/src/main/kotlin/in/tazj/k8s/letsencrypt/acme/Route53Responder.kt
+++ b/src/main/kotlin/in/tazj/k8s/letsencrypt/acme/Route53Responder.kt
@@ -67,11 +67,17 @@ class Route53Responder(val route53: AmazonRoute53) : DnsResponder {
     fun findHostedZone(record: String): Option<HostedZone> {
         val fqdnRecord = determineFqdnRecord(record)
         // Attempt to find the correct hosted zone by matching the longest matching suffix.
-        val matchingZone = route53.listHostedZones().hostedZones
+        val zones = route53.listHostedZones().hostedZones
                 // Filter out private zones to prevent the controller from writing into the wrong zone in a split-brain
                 // DNS setup.
                 .filter { !it.config.isPrivateZone }
                 .filter { fqdnRecord.endsWith(it.name) }
+
+        if (zones.isEmpty()) {
+            return Option.None
+        }
+
+        val matchingZone = zones
                 .reduce { acc, zone ->
                     if (zone.name.length > acc.name.length) {
                         zone

--- a/src/test/kotlin/in/tazj/k8s/letsencrypt/acme/CloudDnsResponderTest.kt
+++ b/src/test/kotlin/in/tazj/k8s/letsencrypt/acme/CloudDnsResponderTest.kt
@@ -35,6 +35,25 @@ class CloudDnsResponderTest {
         Assert.assertEquals("Correct zone is found", "correct", result.get().name)
     }
 
+    @Test
+    fun testNoZonesAvailable() {
+        val iterator = emptyList<Zone>().iterator()
+        val page: Page<Zone> = mock {
+            on { iterateAll() } doReturn (iterator)
+        }
+
+        val dns: Dns = mock {
+            on { listZones() } doReturn (page)
+        }
+
+        val responder = CloudDnsResponder(dns)
+        val testRecord = "_acme-challenge.some.test.tazj.in"
+
+        val result = responder.findMatchingZone(testRecord)
+
+        Assert.assertTrue("No zone is returned", result.isEmpty())
+    }
+
     private fun zoneOf(name: String, dnsName: String): Zone {
         return mock {
             on { getName() } doReturn (name)

--- a/src/test/kotlin/in/tazj/k8s/letsencrypt/acme/Route53ResponderTest.kt
+++ b/src/test/kotlin/in/tazj/k8s/letsencrypt/acme/Route53ResponderTest.kt
@@ -58,4 +58,18 @@ class Route53ResponderTest {
         Assert.assertTrue("A hosted zone is found", result.isDefined())
         Assert.assertEquals("The selected hosted zone is correct", "public", result.get().id)
     }
+
+    @Test
+    fun testNoZonesAvailable() {
+        val testList = emptyList<HostedZone>()
+        val client: AmazonRoute53Client = mock {
+            on { listHostedZones() } doReturn (ListHostedZonesResult().withHostedZones(testList))
+        }
+
+        val responder = Route53Responder(client)
+        val testRecord = "_acme-challenge.some.test.tazj.in"
+        val result = responder.findHostedZone(testRecord)
+
+        Assert.assertTrue("No hosted zone is found", result.isEmpty())
+    }
 }

